---
title: "clark_output_visual"
author: "Rachel Xu"
date: "6/18/2020"
output: html_document
---

**PATH** to clark data directory:
```{r}
path <- "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"
```

**setwd** and Rmd settings
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = path)
```

**Libraries** to include:
```{r}
library(dplyr)
library(ggplot2)
library(tidyverse)
library(plotly)
library(magrittr)


```

Read **phylum** Taxa count into _phylum_full_table_
```{r}
phylum_path <- paste(path, "phylum/" , sep="")
files <- list.files(phylum_path)



i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(phylum_path, file,sep=""), header = TRUE)
  taxa_count <- sample_record %>% select(c("Name","Count")) #%>% subset(Count >= 100)
  colnames(taxa_count) <- c("Name","Count")
  taxa_count %<>% subset(Name != "Chordata" & Name != "UNKNOWN")
  rownames(taxa_count) <- NULL
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    phylum_full_table <- taxa_count
  } else {
    phylum_full_table <- full_join(phylum_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

phylum_full_table[is.na(phylum_full_table)] <- 0


sample_keys <- colnames(phylum_full_table)[-1]
phylum_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")
# phylum_full_table

order <- unique(phylum_full_table$Samples)
phylum_full_table$Samples <- factor(phylum_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))

phylum_full_table %>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.001) -> major_phylum
```

```{r}

phylum_plot <- ggplot(major_phylum, aes(x=Samples,y=Read_Count, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"),plot.caption = element_text(hjust = 0, face= "italic"))+
  labs(title = "Metagenomic Reads Classification with Clark(non-Rat) DB (Phylum)", y= "Number of Reads", x="Samples" , caption= "*For the purpose of visualization, only taxa classified with more than 1% in composition will be shown") 

phylum_plot
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_std_phylum.png")


########Relative Plot#################


phylum_relative <- ggplot(major_phylum, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"),plot.caption = element_text(hjust = 0, face= "italic"))+
  labs(title = "Metagenomic Relative Abundancen with Clark(non-Rat) DB (Phylum)", y= "Percentage", x="Samples" , caption= "*For the purpose of visualization, only taxa classified with more than 1% in composition will be shown")+scale_y_continuous(labels = scales::percent_format()) 

phylum_relative
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_std_phylum_rel.png")
```

Read **genus** Taxa count into _genus_full_table_
```{r}
genus_path <- paste(path, "genus/" , sep="")
files <- list.files(genus_path)

i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(genus_path, file,sep=""))
  taxa_count <- sample_record %>% select(c("Name","Count"))
  colnames(taxa_count) <- c("Name","Count")
  taxa_count %<>% 
    subset(Name != "Homo" & Name != "UNKNOWN") #%>% subset(Count >= 300) 
  
  rownames(taxa_count) <- NULL
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0


sample_keys <- colnames(genus_full_table)[-1]
genus_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")

order <- unique(genus_full_table$Samples)
genus_full_table$Samples <- factor(genus_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))

genus_full_table %>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.01) -> major_genus
```

```{r}

genus_plot <- ggplot(major_genus, aes(x=Samples,y=Read_Count, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.key.size = unit(0.4,"cm"), legend.text = element_text(size = 8),plot.caption = element_text(hjust = 0, face= "italic")) +
  labs(title = "Metagenomic Reads Classification with Clark(non-Rat) DB (Genus)", y= "Number of Reads",
       x="Samples", caption= "*For the purpose of visualization, only taxa classified with more than 1% in composition will be shown")

genus_plot

# ggplotly(genus_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_std_genus.png")

########Relative Plot#################


genus_relative <- ggplot(major_genus, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"),plot.caption = element_text(hjust = 0, face= "italic"))+
  labs(title = "Metagenomic Relative Abundancen with Clark(non-Rat) DB (Genus)", y= "Percentage", x="Samples" , caption= "*For the purpose of visualization, only taxa classified with more than 1% in composition will be shown")+scale_y_continuous(labels = scales::percent_format()) 

genus_relative
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_std_genus_rel.png")
```

Get **Domain** classification from genus results
```{r}

genus_path <- paste(path, "genus/" , sep="")
files <- list.files(genus_path)

i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(genus_path, file,sep=""))
  taxa_count <- sample_record %>% group_by(Lineage) %>% mutate(Domain = sum(Count)) %>% select("Lineage","Domain")
  taxa_count <- unique(taxa_count)

  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    domain_full_table <- taxa_count
  } else {
    domain_full_table <- full_join(domain_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

# domain_full_table
key_sample <- colnames(domain_full_table)[-1]
domain_full_table %<>% gather(key_sample, key = Samples, value = Domain_counts)
order <- c("UNKNOWN", "Eukaryota","Bacteria","Archaea","Viruses")
domain_full_table$Name <- factor(domain_full_table$Name, levels = order)
domain_full_table$Samples <- factor(domain_full_table$Samples, levels= c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))

domain_full_table%>% group_by(Samples) %>%
  mutate(Total = sum(Domain_counts), Percentage = Domain_counts/Total) %>%
  subset(Percentage >= 0.01) -> major_domain
major_domain
```

```{r}

plot_domain <- ggplot(major_domain, aes(x=Samples, y=Domain_counts, fill=Name)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_text(angle=90)) +
  labs(title = "Metagenomic Reads Classification with Clark(non-Rat) DB (Domain)", y= "Number of Reads",
       x="Samples")
plot_domain

# ggplotly(plot_domain)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark_std_domain.png")
########Relative Plot#################

domain_relative <- ggplot(major_domain, aes(x=Samples, y=Percentage, fill=Name)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_text(angle=90)) +
  labs(title = "Metagenomic Relative Abundance Classification with Clark(non-Rat) DB (Domain)",
       x="Samples") +  scale_y_continuous(labels = scales::percent_format()) 
domain_relative
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark_std_domain_rel.png")
```