---
title: "clarks_visual"
output: html_document
---

```{r setup, include=FALSE}
custom_path<- "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = custom_path)

```
###include a color palette for ggplot
```{r}
manual_color <-c("#4361EE","#FFADAD","#FDFFB6","#FFD6A5","#CAFFBF","#A0C4FF","#9BF6FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C","#F94144","#80B918","#FEE440")

```
# Libraries imported for analysis
```{r}

library(dplyr)
library(magrittr)
library(tidyverse)
library(plotly)


```

### since only species level classification has been done for clark (with rat analysis), we will extract Domain information with clark's species results

```{r}

species_path <- paste(custom_path, "species_rat_spaced/" , sep="")
files <- list.files(species_path)



i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(Lineage) %>% mutate(Domain = sum(Count)) %>% select("Lineage","Domain")
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  # taxa_count %<>% subset(Name != "Chordata" & Name != "UNKNOWN")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    domain_full_table <- taxa_count
  } else {
    domain_full_table <- full_join(domain_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

domain_full_table[is.na(domain_full_table)] <- 0
domain_full_table
 
sample_keys <- colnames(domain_full_table)[-1]
domain_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")

order <- unique(domain_full_table$Samples)
domain_full_table$Samples <- factor(domain_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))



domain_full_table$Name <- factor(domain_full_table$Name, levels = c("UNKNOWN", "Eukaryota","Bacteria","Viruses","Archaea"))

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_full_table %>% group_by(Samples) %>% summarise(total_reads=sum(Read_Count)) -> clark_total_reads
as.data.frame(list(mean_reads=mean(clark_total_reads$total_reads),std=sd(clark_total_reads$total_reads))) -> clark_total_reads_summary # avg number of reads across sample
clark_total_reads_summary
```

# Plot Domain level taxaonomy count with UNKNOWN and Eukaryota to have an overall picture of the samples
```{r}
plot_domain <- ggplot(domain_full_table, aes(x=Samples, y=Read_Count, fill=Name)) +
  geom_bar(stat="identity") + 
 theme_classic()+
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position = "none") +
  scale_fill_manual(values = manual_color)
ggplotly(plot_domain)
plot_domain


ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark-s_cus_domain.png",width = 8, height = 6)

#### plot relative abundance 
domain_full_table %>%
  group_by(Samples) %>%
  mutate(Total=sum(Read_Count)) %>%
  mutate(Relative=Read_Count/Total) -> domain_major
########Relative Plot#################

domain_relative <- ggplot(domain_major, aes(x=Samples, y=Relative, fill=Name)) +
  geom_bar(stat="identity") + 
 theme_classic()+
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position = "none") +
  scale_fill_manual(values = manual_color) +
  scale_y_continuous(labels = scales::percent_format(),position = "right") 
ggplotly(domain_relative)
domain_relative
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark-s_cus_domain_rel.png",width = 8, height = 6)


```
# Plot Phylum level abundance without Eukaryota and Unknown information for better visualization 
```{r}
species_path <- paste(custom_path, "species_rat_spaced/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(X) %>% mutate(Phylum = sum(Count)) %>% select("X","Phylum") #%>% subset(Phylum >= 100)
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  taxa_count %<>% subset(Name != "Chordata" & Name != "")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    phylum_full_table <- taxa_count
  } else {
    phylum_full_table <- full_join(phylum_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

phylum_full_table[is.na(phylum_full_table)] <- 0
phylum_full_table
 
sample_keys <- colnames(phylum_full_table)[-1]
phylum_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")

all_phylums <- unique(phylum_full_table$Name) # all phylums classified
# phylums I want to factored into order so they can be assigned with a specific color during plotting



order_phylum <- c("Proteobacteria","Actinobacteria","Cyanobacteria","Firmicutes","Bacteroidetes","Uroviricota","Tenericutes","Spirochaetes","Aquificae","Artverviricota","Peploviricota","Deinococcus-Thermus","Planctomycetes","Gemmatimonadetes","Verrucomicrobia") 

remaining_phylum <- all_phylums %>% data.frame() %>% subset(!c(all_phylums %in% order_phylum))
remaining_phylum <- remaining_phylum$.

phylum_full_table$Samples <- factor(phylum_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
phylum_full_table$Name <- factor(phylum_full_table$Name, level=c(order_phylum,remaining_phylum))

phylum_full_table %>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.001)-> phylum_major

phylum_major %>% subset(Samples== "R22_K" | Samples=="R26_K" | Samples == "R27_K") -> clark_kidney_phylum
unique(clark_kidney_phylum$Name)
```

# Plot Domain level taxaonomy count with UNKNOWN and Eukaryota to have an overall picture of the samples
```{r}
# color palette adjusted to match with Kraken2 color code for genus taxa
manual_color <-c("#4361EE","#9BF6FF","#FFADAD","#FDFFB6","#FFD6A5","#CAFFBF","#A0C4FF","#BDB2FF","#FE7F2D","#619B8A","#EE6352","#59CD90","#3FA7D6","#50514F","#F25F5C","#FAC05E","#A94303","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C","#F94144","#80B918","#FEE440")

plot_phylum <- ggplot(phylum_major, aes(x=Samples, y=Read_Count, fill=Name)) +
  geom_bar(stat="identity") + 
theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position = "none") +
  scale_fill_manual(values = manual_color)

plot_phylum

# ggplotly(plot_phylum)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark-s_cus_phylum.png",width = 8, height =6)

########Relative Plot#################


phylum_relative <- ggplot(phylum_major, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position = "none") +
  scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values = manual_color)

phylum_relative
ggplotly(phylum_relative)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark-s_cus_phylum_rel.png", width = 8, height =6)

```
### genus level classification 
```{r}
species_path <- paste(custom_path, "species_rat_spaced/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record %>% group_by(X.4) %>% mutate(Genus = sum(Count)) %>% select("X.4","Genus") %>% subset(Genus >= 300)
  taxa_count <- unique(taxa_count)
  colnames(taxa_count) <- c("Name","Count")
  
  taxa_count %<>% subset(Name != "Homo" & Name != "Rattus" & Name != "")
  rownames(taxa_count) <- NULL # reset row number
  colnames(taxa_count) <- c("Name", sample_name)
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = "Name")
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0
genus_full_table
 
sample_keys <- colnames(genus_full_table)[-1]
genus_full_table %<>% gather(sample_keys,key="Samples", value="Read_Count")


# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_full_table$Name) # allgenus classified
order_genus <- c("Mycoplasma","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Alistipes","Anabaena","Faecalibacterium","Prevotella") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

order <- unique(genus_full_table$Samples)
genus_full_table$Samples <- factor(genus_full_table$Samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_full_table$Name <- factor(genus_full_table$Name, level=c(order_genus,remaining_genus))

genus_full_table%>% group_by(Samples) %>% 
  mutate(Total = sum(Read_Count), Percentage = Read_Count/Total) %>%
  subset(Percentage >= 0.01)->major_genus

```
### Plot Genus level absolute abundance
```{r}


manual_color <-c("#FDFFB6","#FE7F2D","#FCCA46","#A1C181","#619B8A","#EE6352","#59CD90","#3FA7D6","#50514F","#FFE066","#70C1B3","#C8A439","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff","#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

plot_genus <- ggplot(major_genus, aes(x=Samples, y=Read_Count, fill=Name)) +
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_text(angle=90)) +
  labs(title = "Metagenomic Reads Classification with Clark-S DB (Genus)", y= "Number of Reads",
       x="Samples") +
  theme(legend.text = element_text(size =8 ),legend.key.size =unit(0.5,"cm"))+
  scale_fill_manual(values=manual_color)

plot_genus

ggplotly(plot_genus)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Absolute/clark-s_cus_genus.png")

########Relative Plot#################


genus_relative <- ggplot(major_genus, aes(x=Samples,y=Percentage, fill=Name)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"),plot.caption = element_text(hjust = 0, face= "italic"))+
  labs(title = "Metagenomic Relative Abundancen with Clark-S DB (Genus)", y= "Percentage", x="Samples" , caption= "*For the purpose of visualization, only taxa classified with more than 10% in composition will be shown")+scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values = manual_color)

genus_relative
# ggplotly(phylum_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/plots/Relative/clark-s_cus_genus_rel.png")


```

---
# Downstream statistical Analysis
```{r}

library("phyloseq")
```

## read in all the clarks sample's data
```{r}
custom_path="/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"
species_path <- paste(custom_path, "species_rat_spaced/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record  %>% select("Lineage","X","X.4","Count") %>% group_by(X.4) %>% mutate(Reads=sum(Count)) %>% select(!c(Count))
  taxa_count <- unique(taxa_count)

  colnames(taxa_count) <- c("Domain","Phylum","Genus",sample_name)
  
  taxa_count %<>% subset(Genus != "Homo" & Genus != "Rattus" & Genus != "")
  rownames(taxa_count) <- NULL # reset row number
 
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = c("Domain","Phylum" ,"Genus"))
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0
genus_full_table
```

# put clarks data into phyloseq object
```{r}
# OTU table
otu_data <- genus_full_table %>% select("Genus","R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S") %>% data.frame
rownames(otu_data)<-genus_full_table$Genus
otu_data <- otu_data[,-1]
OTU <- otu_table(otu_data,taxa_are_rows = T)

# lineage table
tax_data <- genus_full_table %>% 
  select("Domain","Phylum","Genus") %>% data.frame
rownames(tax_data) <- tax_data$Genus
TAX <- tax_table(as.matrix(tax_data))

# add sample's metadata
meta_sample <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken2_sample_meta.csv")
row.names(meta_sample) <- meta_sample$X
meta_sample <- meta_sample[,-1]
sample_meta <- sample_data(meta_sample)
sample_meta

# merge all information together
cus_clarks_physeq <- phyloseq(OTU,TAX,sample_meta)
```

### Alpha Diversity (Within Sample diversity)
- tutorial: https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html

- **Shannon**: How difficult it is to predict the identity of a randomly chosen individual.
- **Simpson**: The probability that two randomly chosen individuals are the same species.
- **Inverse Simpson**: This is a bit confusing to think about. Assuming a theoretically community where all species were equally abundant, this would be the number of species needed to have the same Simpson index value for the community being analyzed.

```{r}

library(vegan)

# margin = 2, means samples are the col names
alpha_diversity <- data.frame(Shannon = diversity(otu_data, MARGIN =2, index="shannon"), Simpson = diversity(otu_data, MARGIN =2, index="simpson"), Inverse_Simpson = diversity(otu_data, MARGIN =2, index="invsimpson")) 
alpha_diversity
# plot Alpha diversity 
plot_richness(cus_clarks_physeq, x= "Tissue", color="Subject",measures = c("Shannon", "Simpson", "InvSimpson"),title = "Alpha diversity for each sample")

```

###Beta Diversity (between samples)
https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html
- Bray–Curtis: The sum of lesser counts for species present in both communities divided by the sum of all counts in both communities. This can be thought of as a quantitative version of the Sørensen index.
- Weighted Unifrac: The fraction of the phylogenetic tree branch lengths shared by the two communities, weighted by the counts of organisms, so more abundant organisms have a greater influence.
```{r}

# no margin option like alpha diversity, use transpose for samples as colnames
# bray-curtis
beta_dist <-vegdist(t(otu_data), index="bray") # pairwise comparison between samples

# plot beta diversity using ordination, a way to display high dimensional data
# capture the information in many dimensions by in a smaller number of "artifical" dimensions
mds <- metaMDS(beta_dist)
mds_clarks <- as.data.frame(mds$points)
mds_clarks$samples <- rownames(mds_clarks)
meta_sample$samples <- rownames(meta_sample)
mds_clarks_meta <- left_join(meta_sample,mds_clarks, by="samples")

ggplot(mds_clarks_meta,aes(x=MDS1,y=MDS2,color=Subject))+
  geom_point()
```

### Differential Abundant with Deseq2
- load DESeq2 package
```{r}
library("DESeq2")

```

- differentially abundant analysis with deseq2
```{r}
desseq_clarks <- phyloseq_to_deseq2(cus_clarks_physeq, ~ Tissue)
desseq_clarks
clarks_deseq <- DESeq(desseq_clarks, test="Wald", fitType="parametric")

# pairwise comparison between 2 tissues
Kidney_lung <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Lung"))
kidney_spleen <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Spleen"))
lung_spleen <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Lung","Spleen"))

alpha = 0.01 # significance threhold

# change pairwise comparison results for comparing two different tissues
sigtax_clarks <- Kidney_lung[which(Kidney_lung$padj < alpha),] # get tax below the significance thresholf
sigtax_clarks$Genus <- rownames(sigtax_clarks) # add a column so genus tax can combine with upper lineages

# add lineages to significant taxa
clarks_deseq_lineage <- left_join(as.data.frame(sigtax_clarks), tax_data)
dim(clarks_deseq_lineage)
unique(clarks_deseq_lineage$Phylum)


```

- Plot significantly abundant taxa from the samples
```{r}
# find the maximum log2fold change from deseq2 output for each cell (tapply)
x<-tapply(clarks_deseq_lineage$log2FoldChange,clarks_deseq_lineage$Phylum,function(x)max(x))
x<- sort(x,TRUE) # sort these identified phyla in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clarks_deseq_lineage$Phylum <- factor(as.character(clarks_deseq_lineage$Phylum),levels = names(x))

# perform the same ordering procedure with genus too
x<-tapply(clarks_deseq_lineage$log2FoldChange,clarks_deseq_lineage$Genus,function(x)max(x))
x<- sort(x,TRUE) # sort these identified genus in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clarks_deseq_lineage$Genus <- factor(as.character(clarks_deseq_lineage$Genus),levels = names(x))

ggplot(clarks_deseq_lineage, aes(x=Genus, y=log2FoldChange, color= Phylum))+ geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```